src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -57,49 +57,67 @@ public DiagnosticService([ImportMany] IEnumerable<Lazy<IAsynchronousOperationLis
 
         private void RaiseDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs args)
         {
+            Contract.ThrowIfNull(sender);
+            var source = (IDiagnosticUpdateSource)sender;
+
             var ev = _eventMap.GetEventHandlers<EventHandler<DiagnosticsUpdatedArgs>>(DiagnosticsUpdatedEventName);
-            if (ev.HasHandlers)
+            if (!RequireRunningEventTasks(source, ev))
```

Comment:
Ok... first, where are all these diagnostic sources listed in the code? :)

Second, i've found: IDiagnosticUpdateSource

And i can see how (sorta) how it works.  But it seems like these shoudl really be two different interfaces.  Probably with one interface deriving from the other.  i.e.

```c#
// Simplest form of a diagnostics source.  does nothing but fire events when diagnostics are created. 
// Roslyn does all the book keeping ourselves.

interface IDiagnosticUpdateSource {
        /// <summary>
        /// Raise this when new diagnostics are found
        /// </summary>
        event EventHandler<DiagnosticsUpdatedArgs> DiagnosticsUpdated;
}

interface ICachingDiagnosticsUpdateSource : IDiagnosticsUpdateSource
{
    ImmutableArray<DiagnosticData> GetDiagnostics(Workspace workspace, ProjectId projectId, DocumentId documentId, object id, bool includeSuppressedDiagnostics, CancellationToken cancellationToken);
}
```

Note: it's still not clear why anyone would ever implement the second interface.  When would someone want to have to manage all this themselves instead of just letting our diagnostics engine manage it for them?